crates/swc_html_codegen/src/lib.rs

```diff
@@ -43,6 +43,154 @@ where
+
+        if let Some(name) = &n.name {
+            doctype.push(' ');
+            doctype.push_str(name);
+        }
+
+        if let Some(public_id) = &n.public_id {
+            doctype.push(' ');
+            doctype.push_str("PUBLIC");
+            doctype.push(' ');
+            doctype.push('"');
+            doctype.push_str(public_id);
+            doctype.push('"');
+
+            if let Some(system_id) = &n.system_id {
+                doctype.push(' ');
+                doctype.push('"');
+                doctype.push_str(system_id);
+                doctype.push('"');
+            }
+        } else if let Some(system_id) = &n.system_id {
+            doctype.push(' ');
+            doctype.push_str("SYSTEM");
+            doctype.push(' ');
+            doctype.push('"');
+            doctype.push_str(system_id);
+            doctype.push('"');
+        }
+
+        doctype.push('>');
+
+        write_raw!(self, n.span, &doctype);
+    }
+
+    #[emitter]
+    fn emit_element(&mut self, n: &Element) -> Result {
+        let mut start_tag = String::new();
+
+        start_tag.push('<');
+        start_tag.push_str(&n.tag_name);
+
+        for attribute in &n.attributes {
+            start_tag.push(' ');
+            start_tag.push_str(&attribute.name);
+
+            if let Some(value) = &attribute.value {
+                start_tag.push('=');
+
+                let quote = if value.contains('"') { '\'' } else { '"' };
+
+                start_tag.push(quote);
+                start_tag.push_str(value);
+                start_tag.push(quote);
+            }
+        }
+
+        start_tag.push('>');
+
+        write_str!(self, n.span, &start_tag);
+
+        let no_children = n.namespace == Namespace::HTML
+            && matches!(
+                &*n.tag_name,
+                "area"
+                    | "base"
+                    | "basefont"
+                    | "bgsound"
+                    | "br"
+                    | "col"
+                    | "embed"
+                    | "frame"
+                    | "hr"
+                    | "img"
+                    | "input"
+                    | "keygen"
+                    | "link"
+                    | "meta"
+                    | "param"
+                    | "source"
+                    | "track"
+                    | "wbr"
+            );
+
+        if no_children {
+            return Ok(());
+        }
+
+        if !n.children.is_empty() {
+            self.emit_list(&n.children, ListFormat::NotDelimited)?;
+        }
+
+        let mut end_tag = String::new();
+
+        end_tag.push('<');
+        end_tag.push('/');
+        end_tag.push_str(&n.tag_name);
+        end_tag.push('>');
+
+        write_str!(self, n.span, &end_tag);
+    }
+
+    #[emitter]
+    fn emit_text(&mut self, n: &Text) -> Result {
+        let mut text = String::new();
+
+        for c in n.value.chars() {
+            let new_value = match c {
```

Comment:
I think this line is inefficient, as we don't need allocation here.
We can directly push to `text`